File: app/src/main/java/io/shiny/telega/AuthActivity.java
package io.shiny.telega;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;

import org.drinkless.td.libcore.telegram.Client;
import org.drinkless.td.libcore.telegram.TdApi;

import static io.shiny.telega.Persist.client;

public class AuthActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_auth);
    }

    /*private static void onAuthorizationStateUpdated(TdApi.AuthorizationState authorizationState) {
        if (authorizationState != null) {
            Example.authorizationState = authorizationState;
        }
        switch (Example.authorizationState.getConstructor()) {
            case TdApi.AuthorizationStateWaitTdlibParameters.CONSTRUCTOR:
                TdApi.TdlibParameters parameters = new TdApi.TdlibParameters();
                parameters.databaseDirectory = "tdlib";
                parameters.useMessageDatabase = true;
                parameters.useSecretChats = true;
                parameters.apiId = 94575;
                parameters.apiHash = "a3406de8d171bb422bb6ddf3bbd800e2";
                parameters.systemLanguageCode = "en";
                parameters.deviceModel = Build.MODEL;
                parameters.applicationVersion = "1.0";
                parameters.enableStorageOptimizer = true;

                client.send(new TdApi.SetTdlibParameters(parameters), new AuthorizationRequestHandler());
                break;
            case TdApi.AuthorizationStateWaitEncryptionKey.CONSTRUCTOR:
                client.send(new TdApi.CheckDatabaseEncryptionKey(), new AuthorizationRequestHandler());
                break;
            case TdApi.AuthorizationStateWaitPhoneNumber.CONSTRUCTOR: {
                String phoneNumber = promptString("Please enter phone number: ");
                client.send(new TdApi.SetAuthenticationPhoneNumber(phoneNumber, null), new AuthorizationRequestHandler());
                break;
            }
            case TdApi.AuthorizationStateWaitOtherDeviceConfirmation.CONSTRUCTOR: {
                String link = ((TdApi.AuthorizationStateWaitOtherDeviceConfirmation) Example.authorizationState).link;
                System.out.println("Please confirm this login link on another device: " + link);
                break;
            }
            case TdApi.AuthorizationStateWaitCode.CONSTRUCTOR: {
                String code = promptString("Please enter authentication code: ");
                client.send(new TdApi.CheckAuthenticationCode(code), new AuthorizationRequestHandler());
                break;
            }
            case TdApi.AuthorizationStateWaitRegistration.CONSTRUCTOR: {
                String firstName = promptString("Please enter your first name: ");
                String lastName = promptString("Please enter your last name: ");
                client.send(new TdApi.RegisterUser(firstName, lastName), new AuthorizationRequestHandler());
                break;
            }
            case TdApi.AuthorizationStateWaitPassword.CONSTRUCTOR: {
                String password = promptString("Please enter password: ");
                client.send(new TdApi.CheckAuthenticationPassword(password), new AuthorizationRequestHandler());
                break;
            }
            case TdApi.AuthorizationStateReady.CONSTRUCTOR:
                Persist.hasAuth = true;
                authorizationLock.lock();
                try {
                    gotAuthorization.signal();
                } finally {
                    authorizationLock.unlock();
                }
                break;
            case TdApi.AuthorizationStateLoggingOut.CONSTRUCTOR:
                Persist.hasAuth = false;
                print("Logging out");
                break;
            case TdApi.AuthorizationStateClosing.CONSTRUCTOR:
                Persist.hasAuth = false;
                print("Closing");
                break;
            case TdApi.AuthorizationStateClosed.CONSTRUCTOR:
                print("Closed");
                if (!needQuit) {
                    client = Client.create(new UpdateHandler(), null, null); // recreate client after previous has closed
                } else {
                    canQuit = true;
                }
                break;
            default:
                System.err.println("Unsupported authorization state:" + newLine + Example.authorizationState);
        }
    }*/
    private static class AuthorizationRequestHandler implements Client.ResultHandler {
        @Override
        public void onResult(TdApi.Object object) {
            switch (object.getConstructor()) {
                case TdApi.Error.CONSTRUCTOR:
                    Log.e("Radiogram", "Receive an error:" + object);
                    //onAuthorizationStateUpdated(null); // repeat last action
                    break;
                case TdApi.Ok.CONSTRUCTOR:
                    // result is already received through UpdateAuthorizationState, nothing to do
                    break;
                default:
                    Log.e("Radiogram", "Receive wrong response:" + object);
            }
        }
    }
}

File: app/src/main/java/io/shiny/telega/AuthActivity.kt
package io.shiny.telega

import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.transition.Scene
import androidx.transition.Slide
import androidx.transition.Transition
import androidx.transition.TransitionManager
import io.shiny.telega.Persist.client
import org.drinkless.td.libcore.telegram.Client
import org.drinkless.td.libcore.telegram.TdApi.*

class AuthActivity: AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_auth)
    }
    private fun onAuthorizationStateUpdated(authorizationState: AuthorizationState?) {
        if (authorizationState != null) {
            Persist.authorizationState = authorizationState
        }
        when (Persist.authorizationState.constructor) {
            AuthorizationStateWaitTdlibParameters.CONSTRUCTOR -> {
                val parameters = TdlibParameters()
                parameters.databaseDirectory = "tdlib"
                parameters.useMessageDatabase = true
                parameters.useSecretChats = true
                parameters.apiId = 94575
                parameters.apiHash = "a3406de8d171bb422bb6ddf3bbd800e2"
                parameters.systemLanguageCode = "en"
                parameters.deviceModel = Build.MODEL
                parameters.applicationVersion = "dev"
                parameters.enableStorageOptimizer = true
                client.send(SetTdlibParameters(parameters), AuthorizationRequestHandler())
            }
            AuthorizationStateWaitEncryptionKey.CONSTRUCTOR -> client.send(
                CheckDatabaseEncryptionKey(),
                AuthorizationRequestHandler()
            )
            AuthorizationStateWaitPhoneNumber.CONSTRUCTOR -> {
                //val phoneNumber: String = promptString("Please enter phone number: ")
                goTo(R.layout.auth_phone)
                client.send(
                    SetAuthenticationPhoneNumber(phoneNumber, null),
                    AuthorizationRequestHandler()
                )
            }
            AuthorizationStateWaitOtherDeviceConfirmation.CONSTRUCTOR -> {
                val link =
                    (Persist.authorizationState as AuthorizationStateWaitOtherDeviceConfirmation).link
                println("Please confirm this login link on another device: $link")
            }
            AuthorizationStateWaitCode.CONSTRUCTOR -> {
                val code: String = promptString("Please enter authentication code: ")
                client.send(CheckAuthenticationCode(code), AuthorizationRequestHandler())
            }
            AuthorizationStateWaitRegistration.CONSTRUCTOR -> {
                val firstName: String = promptString("Please enter your first name: ")
                val lastName: String = promptString("Please enter your last name: ")
                client.send(
                    RegisterUser(firstName, lastName),
                    AuthorizationRequestHandler()
                )
            }
            AuthorizationStateWaitPassword.CONSTRUCTOR -> {
                val password: String = promptString("Please enter password: ")
                client.send(
                    CheckAuthenticationPassword(password),
                    AuthorizationRequestHandler()
                )
            }
            AuthorizationStateReady.CONSTRUCTOR -> {
                Persist.hasAuth = true
            }
            AuthorizationStateLoggingOut.CONSTRUCTOR -> {
                Persist.hasAuth = false
                print("Logging out")
            }
            AuthorizationStateClosing.CONSTRUCTOR -> {
                Persist.hasAuth = false
                print("Closing")
            }
            AuthorizationStateClosed.CONSTRUCTOR -> {
                print("Closed")
                if (!needQuit) {
                    client = Client.create(
                        UpdateHandler(),
                        null,
                        null
                    ) // recreate client after previous has closed
                }
            }
            else -> Log.e("Radiogram", "Unsupported authorization state:" + Persist.authorizationState)
        }
    }

    private class AuthorizationRequestHandler : Client.ResultHandler {
        override fun onResult(`object`: Object) {
            when (`object`.constructor) {
                Error.CONSTRUCTOR -> Log.e("Radiogram", "Receive an error:$`object`")
                Ok.CONSTRUCTOR -> {
                }
                else -> Log.e("Radiogram", "Receive wrong response:$`object`")
            }
        }
    }

    fun goTo(rid: Int) {
        val sceneRoot: ViewGroup = findViewById(R.id.scene_root)
        val anotherScene: Scene = Scene.getSceneForLayout(sceneRoot, rid, this)
        var transition: Transition = Slide(Gravity.LEFT)
        TransitionManager.go(anotherScene, transition)
    }
}